test/src/queryhandler.spec.ts

```diff
@@ -94,6 +94,130 @@ describe('Query handler tests', function () {
     });
   });
 
+  describe('Text selectors', function () {
+    describe('in Page', function () {
+      it('should query existing element', async () => {
+        const {page} = getTestState();
+
+        await page.setContent('<section>test</section>');
+
+        expect(await page.$('text/test')).toBeTruthy();
+        expect((await page.$$('text/test')).length).toBe(1);
+      });
+      it('should return empty array for non-existing element', async () => {
+        const {page} = getTestState();
+
+        expect(await page.$('text/test')).toBeFalsy();
+        expect((await page.$$('text/test')).length).toBe(0);
+      });
+      it('should return first element', async () => {
+        const {page} = getTestState();
+
+        await page.setContent('<div>a</div><div>b</div>');
+
+        const element = await page.$('text/a');
+        expect(
+          await element?.evaluate(e => {
+            return e.textContent;
+          })
+        ).toBe('a');
+      });
+      it('should return multiple elements', async () => {
+        const {page} = getTestState();
+
+        await page.setContent('<div>a</div><div>a</div>');
+
+        const elements = await page.$$('text/a');
+        expect(elements.length).toBe(2);
+      });
+      it('should pierce shadow DOM', async () => {
+        const {page} = getTestState();
+
+        await page.setContent(`
+        <div>
+          <div shadowroot="open">
```

Comment:
oh this is not a shadow root actually, just a regular div. I guess you meant to write smth like

```
<host-element>
  <template shadowroot="open">
    <div>a</div>
    <div>b</div>
  </template>
</host-element>
```

Reply:
Ah, okay. The problem is firefox fails with
```
<div>
  <template shadowroot="open">
    <div>a</div>
    <div>b</div>
  </template>
</div>
```

Reply:
you can also create a shadow root and a custom element via JS. I think Firefox might not support declarative shadow DOM